src/extension/experimentService/experimentService.ts

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for details.
+
+import * as Configstore from "configstore";
+import * as https from "https";
+import * as vscode from "vscode";
+import { IExperiment } from "./IExperiment";
+import { PromiseUtil } from "../../common/node/promise";
+import { TelemetryHelper } from "../../common/telemetryHelper";
+import { Telemetry } from "../../common/telemetry";
+
+export enum ExperimentStatuses {
+    ENABLED = "enabled",
+    DISABLED = "disabled",
+    FAILED = "failed",
+}
+
+export interface ExperimentConfig {
+    experimentName: string;
+    popCoveragePercent: number;
+    enabled: boolean;
+}
+
+export interface ExperimentParameters extends ExperimentConfig {
+    [key: string]: any;
+    extensionId?: string;
+}
+
+export interface ExperimentResult {
+    resultStatus: ExperimentStatuses;
+    updatedExperimentParameters: ExperimentParameters;
+    error?: Error;
+}
+
+export class ExperimentService implements vscode.Disposable {
+    private static instance: ExperimentService;
+
+    private readonly endpointURL: string;
+    private readonly configName: string;
+    private config: Configstore;
+    private downloadedExperimentsConfig: Array<ExperimentConfig> | null;
+    private experimentsInstances: Map<string, IExperiment>;
+    private downloadConfigRequest: Promise<ExperimentConfig[]>;
+    private promiseUtil: PromiseUtil;
+    private cancellationTokenSource: vscode.CancellationTokenSource;
+
+    public static create () {
+        if (!ExperimentService.instance) {
+            ExperimentService.instance = new ExperimentService();
+        }
+
+        return ExperimentService.instance;
+    }
+
+    public async runExperiments(): Promise<void> {
+        if (!this.downloadedExperimentsConfig) {
+            this.downloadedExperimentsConfig = await this.downloadConfigRequest;
+            this.experimentsInstances = await this.initializeExperimentsInstances();
+        }
+
+        let experimentResults: Array<ExperimentResult> = await Promise.all(this.downloadedExperimentsConfig
+            .map(async (expConfig) => await this.executeExperiment(expConfig))
```

Comment:
I think that you don't need to await in the following line:

```suggestion
            .map(expConfig => this.executeExperiment(expConfig))
```

This should return already a promise that will be "awaited" in `await Promise.all` .